### ISSUE TITLE
Intermediary steps and the "pyramid of doom".

### ISSUE THREAD

**Issue by criske** [self-xdsd/self-core#473]
Having a lot of intermediary steps could lead to this, or at least will make the code somewhat hard to follow.
It's starting to show here already:
https://github.com/self-xdsd/self-core/blob/6e62de8855ac0aabf9a9aca06e73e3589f6f4b6d/self-core-impl/src/main/java/com/selfxdsd/core/managers/Deregister.java#L69
```java
new UnassignTask(
       new RemoveTask(
              new SendReply(String.format(language.reply("deregister.comment"), author),
                lastly -> LOG .debug("Task removed successfully.")
            )
      )
 )
```
We could easly flatten this by using a builder/composer:
```java
new StepComposer()
    .next(UnassignTask::new)
    .next(RemoveTask::new)
    .next(s -> new SendReply(String.format(language.reply("deregister.comment"), author), s))
    .finish(lastly -> LOG.debug("Task removed successfully."))
```

I'll do a quick PR to see how it works, no bounty needed.

**Comment by amihaiemil** [comment]
@criske yes, there can be many steps added. But, on the other hand, this is a case of ``do-and-forget``. We wrote Deregister once and we will probably never touch it again, unless we remove it - it's a core functionality of the PM which will be altered very rarely, I think.

**Comment by amihaiemil** [comment]
@criske And how do we take care of the PreconditionCheck steps? Those steps need 2 arguments: onTrue and onFalse. How will they look in this composer? :D

**Comment by criske** [comment]
@amihaiemil well composer is just to chain Intermediary steps, at least for now :).

**Comment by amihaiemil** [comment]
@criske 
> well composer is just to chain Intermediary step

Most of the time we will have PreconditionCheck steps at the top of the pyramid, so it's pretty useless at the moment. But the overall idea is ok, let's keep it open and we'll come back to it if we really need to.

**Comment by criske** [comment]
@amihaiemil but sure could be adapted like:
```java
 new StepComposer()
  .when(onTrue, onFalse -> PreconditionCheck(onTrue, onFalse))
  .onTrue(s -> new StepComposer()...finish())
  .onFalse(s -> new StepComposer()...finish())
  .finish()
```
